src/tmo/fattal02/pfstmo_fattal02.1
.TH "pfstmo_fattal02" 1
.SH NAME
pfstmo_fattal02 \- Gradient domain high dynamic range compression
.SH SYNOPSIS
.B pfstmo_fattal02
[--alpha <val>] [--beta <val>]
[--gamma <val>]
[--saturation <val>]
[--noise <val>]
[--detail-level <val>]
[--black-point <val>] [--white-point <val>]
[--multigrid]
[--verbose] [--help]
.SH DESCRIPTION
This command implements a tone mapping operator as described in:

Gradient Domain High Dynamic Range Compression
R. Fattal, D. Lischinski, and M. Werman
In ACM Transactions on Graphics, 31(3), p. 249, 2002.

With respect to the original paper, this program provides additional parameter
which limits the amplification of noise. The noise is often starkly amplified
because of division by zero in one of the equations in the paper. Extension
contributed by Przemyslaw Bazarnik.

At the core of the programme is a Poisson PDE which as suggested in
the original paper is solved using a Full Multigrid
Algorithm. However, this is an iterative solver which seems to lose
accuracy when applied to higher resolution images resulting in halo
effects and surreal looking images. For that reason a second solver
has been implemented using the discrete cosine transform as the
underlying method and is considerably more accurate mainly because it
is a direct solver. This solver is the preferred method and is used by
default. The old multigrid solver can be selected with the --multigrid
(-m) option. 
.SH OPTIONS
.TP
--alpha <val>, -a <val>

Set alpha parameter. This parameter is depreciated as setting a <val> other
than 1.0 has only the effect of a global gamma adjustment of the luminance
channel which can be directly specified using the --gamma option. See the
paper for the definition of alpha. It can be shown, although not mentioned in
the paper, that setting alpha other than 1.0 has the same effect as setting
gamma = alpha^(k*(1-beta)), where beta is the value as specified by --beta and
k is the number of levels of the Gaussian Pyramid (see paper for details),
which depends on the image pixel size (smallest k so that
2^(k+detail_level) >= min(width,height)/MSIZE, MSIZE see source code, e.g. 8 or
32).
.TP
--beta <val>, -b <val>

Set beta parameter. <val> sets the strength of gradient (local contrast) modification. 
Suggested range is 0.8 to 0.96, default is 0.9 (see paper for details). 
Value of 1 does not change contrasts, values above 1 reverse the effect: 
local contrast is stretched and details are attenuated.
Values below 0.5 lead to very strong amplification of small contrast, so
consider using --noise parameter to prevent noise.
.TP
--gamma <val>, -g <val>

Set luminance gamma adjustment. This can be described as a global contrast
enhancement and is applied after the local enhancement as specified by the
parameter --beta is performed. Gamma adjustment or correction is defined
by a power-law, in this case
 L_out(x,y) = L_in(x,y)^gamma,
where L_in(x,y)=exp(I(x,y)) is the luminance value after the local contrast
enhancement (I is the solution of the Poisson PDE). The suggested range for
<val> is 0.6 to 1.0, default is 0.8.
.TP
--saturation <val>, -s <val>

Amount of color saturation. Suggested range is 0.4 to 0.8. Default
value: 0.8.
.TP
--noise <val>, -n <val>

Reduces the gradient amplification value for gradients close to 0 and reduces
noise as a result. <val> defines gradient value (luminance difference of 
adjacent pixels) which is treated as noise. Suggested range is 0.0 to the 
value of alpha. Default value calculated based on alpha: 0.001*alpha.
.TP
--detail-level <val>, -d <val>

Specifies up to which detail level the local contrast enhancement should be
performed. It basically means that local contrast levels within small
squares of pixel size 2^<val> are not changed.
In the implementation this corresponds to removing the <val> finest levels of
the Gaussian Pyramid as described in the paper, i.e. the paper only considers
<val>=0. Suggested values are 1, 2 or 3; 3 for high resolution images. The
default is 3 for --fftsolver, and 0 if the original multi-level solver is used
(to be consistent with the paper).

.TP
--white-point <val>, -w <val>

Specifies the percentage of pixels which are allowed to be overexposed and
therefore blown out. This can be useful for example when there is a very bright
object in the image like the sun and details of it do not need to be resolved.
As a result the overall image will look brighter the greater <val> is.
Default is 0.5.
.TP
--black-point <val>, -k <val>

Same as --white-point but for under-exposed pixels. Default is 0.1.
.TP
--multigrid, -m

Enable the use of the multigrid solver as suggested by the original
paper. For accuracy the default fft solver is generally recommended
especially when using high resolution images. The user will benefit by
obtaining photo-realistic rather than surreal looking images. The fft
solver is also faster despite the fact it is only O(n*log n) with
n=width*height, as compared to O(n) for the multigrid solver.  The
speed improvement is thanks to the very efficient fftw3 library which
is used to calculate the discrete cosine transform.

.TP
--verbose

Print additional information during program execution.
.TP
--help

Print list of command line options.
.SH EXAMPLES
.TP
pfsin memorial.hdr | pfstmo_fattal02 -v -t | pfsout memorial.png

Tone map image (using fft solver) and save it in png format.
.TP
pfsin memorial.hdr | pfstmo_fattal02 -v -t -b 0.85 -g 0.7 -w 2.0 \\
| pfsout memorial.png

Tone map image (using fft solver) with stronger contrast modification than
default, i.e. beta=0.85, gamma=0.7 and white point 2.0%. 
.TP
pfsin memorial.hdr | pfstmo_fattal02 -v | pfsout memorial.png

Tone map image (old style) and save it in png format.
.SH "SEE ALSO"
.BR pfsin (1)
.BR pfsout (1)
.BR pfsview (1)
.SH KNOWN ISSUES
For stronger local contrast enhancements (beta<0.9) the fft solver
(--fftsolver) might produce slightly dark image corners.
This can be mitigated using bigger values for the --noise parameter.

With a value of --detail-level greater than 0, the internal implementation
could be made much more efficient as only a reduced sized PDE would need
to be solved, greatly improving speed.
.SH BUGS
Please report bugs and comments on implementation to the pfstools discussion group
(http://groups.google.com/group/pfstools). For bugs specific to the
FFT solver email Tino Kluge <tino.kluge@hrz.tu-chemnitz.de>.


NEWS
This file is not maintained, but required by automake. Visit pfstools
web page to find the latest news:

http://pfstools.sourceforge.net/



src/tmo/fattal02/pfstmo_fattal02.cpp
/**
 * @file pfstmo_fattal02.cpp
 * @brief Tone map XYZ channels using Fattal02 model
 *
 * Gradient Domain High Dynamic Range Compression
 * R. Fattal, D. Lischinski, and M. Werman
 * In ACM Transactions on Graphics, 2002.
 *
 * 
 * This file is a part of PFSTMO package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Grzegorz Krawczyk, <krawczyk@mpi-sb.mpg.de>
 *
 * $Id: pfstmo_fattal02.cpp,v 1.10 2012/06/22 12:01:32 rafm Exp $
 */

#include <config.h>

#include <iostream>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <math.h>

#include <pfs.h>

#include "tmo_fattal02.h"

using namespace std;

#define PROG_NAME "pfstmo_fattal02"

/// verbose mode
bool verbose = false;

class QuietException 
{
};

void printHelp()
{
  fprintf( stderr, PROG_NAME " (" PACKAGE_STRING ") : \n"
    "\t[--alpha <val>] [--beta <val>] \n"
    "\t[--gamma <val>] \n"
    "\t[--saturation <val>] \n"
    "\t[--noise <val>] \n"
    "\t[--detail-level <val>] \n"
    "\t[--black-point <val>] [--white-point <val>] \n"
    "\t[--multigrid] \n"
    "\t[--verbose] [--help] \n"
    "See man page for more information.\n" );
}

void pfstmo_fattal02( int argc, char* argv[] )
{
  pfs::DOMIO pfsio;

  //--- default tone mapping parameters;
  float opt_alpha = 1.0f;
  float opt_beta = 0.9f;
  float opt_gamma = -1.0f;    // not set (0.8 for fft solver, 1.0 otherwise)
  float opt_saturation=0.8f;
  float opt_noise = -1.0f;    // not set 
  int   opt_detail_level=-1;  // not set (2 for fft solver, 0 otherwise)
  float opt_black_point=0.1f;
  float opt_white_point=0.5f;
  bool  opt_fftsolver=true;
  
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "multigrid", no_argument, NULL, 'm' },
    { "alpha", required_argument, NULL, 'a' },
    { "beta", required_argument, NULL, 'b' },
    { "gamma", required_argument, NULL, 'g' },
    { "saturation", required_argument, NULL, 's' },
    { "noise", required_argument, NULL, 'n' },
    { "detail-level", required_argument, NULL, 'd' },
    { "white-point", required_argument, NULL, 'w' },
    { "black-point", required_argument, NULL, 'l' },
    { NULL, 0, NULL, 0 }
  };

  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, "hvma:b:g:s:n:d:w:k:", cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'm':
      opt_fftsolver = false;
      break;
    case 'a':
      opt_alpha = (float)strtod( optarg, NULL );
      if( opt_alpha<=0.0f )
        throw pfs::Exception("alpha value out of range, should be >0");
      break;
    case 'b':
      opt_beta = (float)strtod( optarg, NULL );
      if( opt_beta<=0.0f )
        throw pfs::Exception("beta value out of range, should be >0");
      break;
    case 'g':
      opt_gamma = (float)strtod( optarg, NULL );
      if( opt_gamma<=0.0f || opt_gamma>=10.0 )
        throw pfs::Exception("gamma value out of range, should be >0");
      break;
    case 's':
      opt_saturation = (float)strtod( optarg, NULL );
      if( opt_saturation<=0.0f || opt_saturation>1.0f )
        throw pfs::Exception("saturation value out of range, should be 0..1");
      break;
    case 'n':
      opt_noise = (float)strtod( optarg, NULL );
      if( opt_noise<0.0f )
        throw pfs::Exception("noise level value out of range, should be >=0");
      break;
    case 'd':
      opt_detail_level = (int) strtod( optarg, NULL );
      if( opt_detail_level<0 || opt_detail_level>9 )
        throw pfs::Exception("detail-level value out of range, should be 0..9");
      break;
    case 'w':
      opt_white_point = (float)strtod( optarg, NULL );
      if( opt_white_point<0.0f || opt_white_point>=50.0f )
        throw pfs::Exception("white-point value out of range, should be 0..50");
      break;
    case 'k':
      opt_black_point = (float)strtod( optarg, NULL );
      if( opt_black_point<0.0f || opt_black_point>=50.0f )
        throw pfs::Exception("black-point value out of range, should be 0..50");
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  if( opt_fftsolver ) {
    // Default params for the fftsolver
    if(opt_detail_level==-1)
      opt_detail_level=3;     // 2 for low res, 3 is better for high res
    if(opt_gamma==-1.0f)
      opt_gamma=0.8f;
    if(opt_noise==-1.0f)
      opt_noise=0.002f;
  }

  
  // adjust noise floor if not set by user
  if( opt_noise<0.0f )
    opt_noise = opt_alpha*0.01;
  // set gamma and detail level to default values which produce the
  // same output as older programme versions using the multi-level pde solver
  if(opt_detail_level==-1)
    opt_detail_level=0;
  if(opt_gamma==-1.0f)
    opt_gamma=1.0f;

  VERBOSE_STR << "threshold gradient (alpha): " << opt_alpha << endl;
  VERBOSE_STR << "strengh of modification (beta): " << opt_beta << endl;
  VERBOSE_STR << "gamma: " << opt_gamma << endl;
  VERBOSE_STR << "noise floor: " << opt_noise << endl;  
  VERBOSE_STR << "saturation: " << opt_saturation << endl;
  VERBOSE_STR << "detail level: " << opt_detail_level << endl;
  VERBOSE_STR << "white point: " << opt_white_point << "%" << endl;
  VERBOSE_STR << "black point: " << opt_black_point << "%" << endl;
  VERBOSE_STR << "use fft pde solver: " << opt_fftsolver << endl;

  while( true ) 
  {
    pfs::Frame *frame = pfsio.readFrame( stdin );
    if( frame == NULL ) break; // No more frames

    pfs::Channel *X, *Y, *Z;
    frame->getXYZChannels( X, Y, Z );
    frame->getTags()->setString("LUMINANCE", "RELATIVE");
    //---

    if( Y==NULL || X==NULL || Z==NULL)
      throw pfs::Exception( "Missing X, Y, Z channels in the PFS stream" );
        
    // tone mapping
    int w = Y->getCols();
    int h = Y->getRows();

    pfs::Array2DImpl* L = new pfs::Array2DImpl(w,h);
    tmo_fattal02(w, h, Y->getRawData(), L->getRawData(), opt_alpha, opt_beta,
            opt_gamma, opt_noise, opt_detail_level,
            opt_black_point, opt_white_point, opt_fftsolver);

    // in-place color space transform
    pfs::Array2DImpl *G = new pfs::Array2DImpl( w, h ); // copy for G to preserve Y
    pfs::Array2D *R = X, *B = Z;
    pfs::transformColorSpace( pfs::CS_XYZ, X, Y, Z, pfs::CS_RGB, R, G, B );
    
    // Color correction 
    for( int i=0; i < w*h; i++ )
    {
      const float epsilon = 1e-4f;
      float y = max( (*Y)(i), epsilon );
      float l = max( (*L)(i), epsilon );
      (*R)(i) = powf( max((*R)(i)/y,0.f), opt_saturation ) * l;      
      (*G)(i) = powf( max((*G)(i)/y,0.f), opt_saturation ) * l;
      (*B)(i) = powf( max((*B)(i)/y,0.f), opt_saturation ) * l;
    }
    
    
    pfs::transformColorSpace( pfs::CS_RGB, R, G, B, pfs::CS_XYZ, X, Y, Z );

    delete G;    
    delete L;
    
    //---
    pfsio.writeFrame( frame, stdout );
    pfsio.freeFrame( frame );        
  }
}

int main( int argc, char* argv[] )
{
  try {
    pfstmo_fattal02( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}
